ansible_runner/streaming.py

```diff
@@ -0,0 +1,214 @@
+import base64
+import codecs
+import io
+import json
+import os
+import stat
+import tempfile
+import uuid
+import zipfile
+
+import ansible_runner
+import ansible_runner.plugins
+
+
+class UUIDEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, uuid.UUID):
+            return obj.hex
+        return json.JSONEncoder.default(self, obj)
+
+
+# List of kwargs options to the run method that should be sent to the remote executor.
+remote_run_options = (
+    'forks',
+    'host_pattern',
+    'ident',
+    'ignore_logging',
+    'inventory',
+    'limit',
+    'module',
+    'module_args',
+    'omit_event_data',
+    'only_failed_event_data',
+    'playbook',
+    'verbosity',
+)
+
+
+class StreamController(object):
+    def __init__(self, control_in, control_out, status_handler=None, event_handler=None,
+                 artifacts_handler=None, cancel_callback=None, finished_callback=None, **kwargs):
+        self.control_in = control_in
+        self.control_out = control_out
+
+        self.kwargs = kwargs
+        self.config = ansible_runner.RunnerConfig(**kwargs)
+        self.status_handler = status_handler
+        self.event_handler = event_handler
+        self.artifacts_handler = artifacts_handler
+
+        self.cancel_callback = cancel_callback
+        self.finished_callback = finished_callback
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def run(self):
+        self.send_job()
+
+        job_events_path = os.path.join(self.config.artifact_dir, 'job_events')
+        if not os.path.exists(job_events_path):
+            os.mkdir(job_events_path, 0o700)
+
+        for line in self.control_in:
+            data = json.loads(line)
+            if 'status' in data:
+                self.status_callback(data)
+            elif 'artifacts' in data:
+                self.artifacts_callback(data)
+            elif 'eof' in data:
+                break
+            else:
+                self.event_callback(data)
+
+        if self.finished_callback is not None:
+            self.finished_callback(self)
+        return self.status, self.rc
+
+    def send_job(self):
+        self.config.prepare()
+        remote_options = {key: value for key, value in self.kwargs.items() if key in remote_run_options}
```

Comment:
It seems like it would be easier to maintain a list of options to _exclude_ rather than a list of options to _allow_.

If I understand the problem this is trying to solve, you want to remove all options _related to streaming_. Those options should be a fairly static and definite list, like `control_in`, `control_out`, `worker_in`, etc.

Reply:
> If I understand the problem this is trying to solve, you want to remove all options _related to streaming_. Those options should be a fairly static and definite list, like `control_in`, `control_out`, `worker_in`, etc.

I don't believe that is actually the case.  There are a number of parameters that do not make sense to be user-provided on the worker side of this streaming scenario.  Off the top of my head, any of the callbacks/handlers.

Reply:
Yeah, if someone adds a new option, it's kinda 50/50 whether it will make sense to send to the remote side, and the safe default is not to send it.

Reply:
There are 50 parameters at this interface.

https://github.com/ansible/ansible-runner/blob/b7c247aa224e45eb86c3ce5b79a025e8ac430716/ansible_runner/interface.py#L88-L151

See in `Worker` how it uses `r = ansible_runner.interface.run(**self.kwargs)`, so this is the method I linked to.

You're including 12. I don't think you are correct in omitting the other 38.

To throw a few out there, `fact_cache`, `cmdline`, `process_isolation`...